### PR TITLE
Fixes progression scaling and other various problems with progression traitors.

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

Removes the 'take' option from the traitor panel for traitor uplinks, as if you take an uplink, you can't give back the objectives to a traitor. The 'take' option still exists for other such uplinks that don't have objectives.
Fixes progression diminishing return scaling, it wasn't scaling exponentially because I messed up the math. This fixes that so that it scales more exponentially. These are the values I got from testing:

40 minutes and above is 90% reduction
30 minutes is 60% reduction
20 minutes is 30% reduction
10 minutes and below is no reduction

This means if you are at 120 minutes, when the round is at 90 minutes, you get a 60% reduction reward to all your objectives. If you are at 130 minutes, it's a 90% reduction and so on.
The scaling is exponential, so it'll be noticeable only if you are exceeding the expected progression by a lot.

## Why It's Good For The Game

Fixes a bug, since it was intended for this to scale exponentially, but I messed up the math so it actually did the opposite and scaled more slowly as the value got greater.

## Changelog

:cl:
fix: Fixed how the diminishing returns of having too much progression scaled. It now properly scales exponentially.
del: Removed the 'take' option for traitor uplinks in the traitor panel.
/:cl:

### CODE REVIEW

File: code/modules/admin/antag_panel.dm [hunk 1]
```diff
@@ -186,7 +186,8 @@ GLOBAL_VAR(antag_prototypes)
 		var/uplink_info = "<i><b>Uplink</b></i>:"
 		var/datum/component/uplink/U = find_syndicate_uplink()
 		if(U)
-			uplink_info += "<a href='?src=[REF(src)];common=takeuplink'>take</a>"
+			if(!U.uplink_handler.has_objectives)
+				uplink_info += "<a href='?src=[REF(src)];common=takeuplink'>take</a>"
```
Comment (JohnFulpWillard) [inline code/modules/admin/antag_panel.dm:190]: if you remove someone's antag datum, are you still unable to take this away? Or does removing the antag datum remove the uplink handler's var?

**Comment by Dorsisdwarf** [comment]
Why is the "remove uplink" option being deleted? I'm not sure I understand the stated reason.

**Comment by Watermelon914** [comment]
> Why is the "remove uplink" option being deleted? I'm not sure I understand the stated reason.

traitors have objectives that are assigned by their traitor datum when their uplink gets made. The 'give' option won't give back their objectives and will effectively wipe their progress. The remove uplink option is only being removed for traitors

File: code/modules/admin/antag_panel.dm (hunk 1 above)
Comment (Watermelon914) [reply code/modules/admin/antag_panel.dm:190]: fixed

**Comment by Timberpoes** [comment]
> > Why is the "remove uplink" option being deleted? I'm not sure I understand the stated reason.
> 
> traitors have objectives that are assigned by their traitor datum when their uplink gets made. The 'give' option won't give back their objectives and will effectively wipe their progress. The remove uplink option is only being removed for traitors

I still don't understand why losing progression is a problem when an admin decides to take their uplink away, and such a problem that admins should not be able to take their uplink away.

Is there a way for admins to wipe their objectives so their uplink can be removed without resorting to proc calls/VV/SDQL2?

**Comment by Watermelon914** [comment]
> > > Why is the "remove uplink" option being deleted? I'm not sure I understand the stated reason.
> > 
> > 
> > traitors have objectives that are assigned by their traitor datum when their uplink gets made. The 'give' option won't give back their objectives and will effectively wipe their progress. The remove uplink option is only being removed for traitors
> 
> I still don't understand why losing progression is a problem when an admin decides to take their uplink away, and such a problem that admins should not be able to take their uplink away.
> 
> Is there a way for admins to wipe their objectives so their uplink can be removed without resorting to proc calls/VV/SDQL2?

remove traitor datum, remove uplink, add traitor datum

**Comment by Mothblocks** [comment]
Honestly I think if you just added a confirm prompt to the take button I wouldn't have fucked up what I did, I don't think it's necessary to remove

**Comment by Watermelon914** [comment]
> Honestly I think if you just added a confirm prompt to the take button I wouldn't have fucked up what I did, I don't think it's necessary to remove

it isn't fully removed, you just can't remove the uplink for traitors until you remove the traitor datum itself

**Comment by Bawhoppen** [comment]
Good pr

**Comment by Mothblocks** [comment]
@iprice If what Watermelon says is correct, and it still shows up if they're not a traitor, then what's the problem? I'm confused as I didn't really use this feature much.

**Comment by Watermelon914** [comment]
> @iprice If what Watermelon says is correct, and it still shows up if they're not a traitor, then what's the problem? I'm confused as I didn't really use this feature much.

basically they're saying they want to be able to take a traitor's tc away for tc trades so that they can't complete objectives, but I think locking traitors out of objectives kinda defeats the purpose, and you're better off just removing their traitor datum and giving them a new one if admins really want to do tc trades on that scale

**Comment by iprice** [comment]
Indeed, intention is to lock them out of further gaining TC.  And that's probably a good alternative, there's always "custom antag" as a placeholder marker for freeform antags.  Just an initial reaction to a button going away I had used a couple of times, all good thanks!

**Pull request merged**